File: servers/python/satellite_server/router/route_builder.py
"""FastAPI router builder for the Satellite sync protocol."""

from __future__ import annotations

import asyncio
import re
from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Any, Callable, Awaitable

import httpx

from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse

from satellite_server.interfaces import IObjectStore
from satellite_server.config.schema import SyncConfig, CollectionConfig, SyncTrigger, WildcardRemoteConfig, WriteMode
from satellite_server.encryption.encrypted_store import EncryptedObjectStore
from satellite_server.protocol.pull import pull
from satellite_server.router.helpers import (
    handle_sync_pull,
    handle_sync_push,
    validate_path_segment,
    SignatureVerifier,
)
from satellite_server.router.middleware import check_body_limit, RateLimiter
from satellite_server.constants import (
    ROLE_PUBLIC,
    ROLE_SELF,
    OP_READ,
    OP_WRITE,
    ENCRYPTION_IDENTITY,
    ENCRYPTION_SERVER,
    ENCRYPTION_DELEGATED,
    ACTION_PULL,
    ACTION_PUSH,
    IDENTITY_PARAM,
    IDENTITY_KEY,
    QUERY_CHECKPOINT,
    HKDF_INFO_IDENTITY,
    HKDF_INFO_SERVER,
)

if TYPE_CHECKING:
    from satellite_server.replica.manager import ReplicaManager
    from satellite_server.replica.notifier import NotificationPublisher


@dataclass
class AuthResult:
    """Result of authenticating a request."""

    identity: str
    roles: list[str]


RoleResolver = Callable[[Request], Awaitable[AuthResult]]
RoleEnricher = Callable[[AuthResult, dict[str, str]], Awaitable[list[str]]]


@dataclass
class SyncRouterOptions:
    """Options for creating a sync router."""

    store: IObjectStore
    config: SyncConfig
    role_resolver: RoleResolver
    role_enricher: RoleEnricher | None = None
    encryption_secret: str | None = None
    server_encryption_secret: str | None = None
    server_identity: str | None = None
    identity_encryption_info: str | None = None
    server_encryption_info: str | None = None
    signature_verifier: SignatureVerifier | None = None
    replica_manager: ReplicaManager | None = None
    """Replica-side: enables ``on_pull`` sync trigger and ``write_through`` push proxying."""
    notification_publisher: NotificationPublisher | None = None
    """Primary-side: fan-out push notifications to subscribed replicas after each write."""


# ── Helpers ──────────────────────────────────────────────────────────────


def _to_route_path(action: str, storage_path: str) -> str:
    """Convert config storagePath template to a FastAPI route path.

    "users/{identity}/invoices" -> "/pull/users/{identity}/invoices"
    """
    return f"/{action}/{storage_path}"


def _resolve_document_key(template: str, params: dict[str, str]) -> str:
    """Resolve a storage path template into a document key."""
    result = template
    for key, value in params.items():
        result = result.replace(f"{{{key}}}", value)
    return result


def _validate_all_params(params: dict[str, str]) -> bool:
    for value in params.values():
        if not validate_path_segment(value):
            return False
    return True


def _extract_path_params(storage_path: str, request_path: str, action: str) -> dict[str, str]:
    """Extract path parameters from a request path using the storage path template."""
    param_names = re.findall(r"\{(\w+)\}", storage_path)
    pattern_str = storage_path
    for name in param_names:
        pattern_str = pattern_str.replace(f"{{{name}}}", f"(?P<{name}>[^/]+)")
    prefix = f"/{action}/"
    path_after_prefix = request_path[len(prefix):] if request_path.startswith(prefix) else request_path
    match = re.match(pattern_str, path_after_prefix)
    if not match:
        return {}
    return match.groupdict()


# ── Auth ─────────────────────────────────────────────────────────────────


async def _check_auth(
    col: CollectionConfig,
    operation: str,
    request: Request,
    params: dict[str, str],
    opts: SyncRouterOptions,
) -> tuple[str | None, JSONResponse | None]:
    """Check authorization. Returns (identity, error_response)."""
    required_roles = col.read_roles if operation == OP_READ else col.write_roles

    if ROLE_PUBLIC in required_roles:
        return None, None

    try:
        auth = await opts.role_resolver(request)
    except Exception:
        return None, JSONResponse({"error": "Unauthorized"}, status_code=401)

    effective_roles = set(auth.roles)

    # Auto-grant "self" when {identity} in path matches authenticated identity
    if IDENTITY_PARAM in col.storage_path:
        if params.get(IDENTITY_KEY) == auth.identity:
            effective_roles.add(ROLE_SELF)

    # Enrich roles
    if opts.role_enricher:
        extra = await opts.role_enricher(auth, params)
        effective_roles.update(extra)

    # Check access
    has_access = any(r in effective_roles for r in required_roles)
    if not has_access:
        return auth.identity, JSONResponse({"error": "Forbidden"}, status_code=403)

    return auth.identity, None


# ── Store resolution ─────────────────────────────────────────────────────


def _resolve_store(
    col: CollectionConfig,
    base_store: IObjectStore,
    params: dict[str, str],
    identity: str | None,
    opts: SyncRouterOptions,
) -> IObjectStore:
    if col.encryption == ENCRYPTION_IDENTITY:
        if not opts.encryption_secret:
            raise RuntimeError(f'Collection "{col.name}" requires encryption_secret')
        salt = identity or params.get(IDENTITY_KEY, "")
        return EncryptedObjectStore(
            base_store,
            opts.encryption_secret,
            salt,
            opts.identity_encryption_info or HKDF_INFO_IDENTITY,
        )
    if col.encryption == ENCRYPTION_SERVER:
        if not opts.server_encryption_secret:
            raise RuntimeError(f'Collection "{col.name}" requires server_encryption_secret')
        if not opts.server_identity:
            raise RuntimeError(f'Collection "{col.name}" requires server_identity')
        return EncryptedObjectStore(
            base_store,
            opts.server_encryption_secret,
            opts.server_identity,
            opts.server_encryption_info or HKDF_INFO_SERVER,
        )
    # ENCRYPTION_NONE and ENCRYPTION_DELEGATED: no server-side encryption.
    # Delegated encryption is handled entirely client-side.
    return base_store


# ── Write-through proxy ───────────────────────────────────────────────────


async def _proxy_push_to_primary(
    col: CollectionConfig,
    request: Request,
    replica_manager: ReplicaManager,
) -> JSONResponse:
    """Forward a push request to the primary satellite server (write_through mode).

    Returns the primary's response directly to the client, then triggers a
    background sync so the local store is updated with the primary's final state.
    """
    remote = col.remote  # type: ignore[union-attr]
    primary_url = f"{remote.url.rstrip('/')}{remote.push_path}"

    raw_body = await request.body()
    headers = {
        "Content-Type": "application/json",
        "Accept": "application/json",
        **remote.headers,
    }

    async with httpx.AsyncClient(timeout=30.0) as client:
        try:
            resp = await client.post(primary_url, content=raw_body, headers=headers)
        except Exception as exc:  # noqa: BLE001
            return JSONResponse(
                {"error": f"Failed to reach primary: {exc}"},
                status_code=502,
            )

    # Mirror the primary's status code to the client
    if resp.status_code == 409:
        return JSONResponse({"error": "hash_mismatch"}, status_code=409)
    if not resp.is_success:
        return JSONResponse(
            {"error": f"Primary returned {resp.status_code}"},
            status_code=resp.status_code,
        )

    # Sync local store from primary in background
    asyncio.create_task(replica_manager.sync_now(col.name))

    return JSONResponse(resp.json(), status_code=resp.status_code)


# ── Route building ───────────────────────────────────────────────────────


def _add_collection_routes(
    router: APIRouter,
    col: CollectionConfig,
    opts: SyncRouterOptions,
) -> None:
    # Pull route
    if not col.push_only:
        pull_path = _to_route_path(ACTION_PULL, col.storage_path)

        async def pull_handler(request: Request, col=col) -> JSONResponse:
            params = request.path_params
            if not _validate_all_params(params):
                return JSONResponse({"error": "Invalid path parameter"}, status_code=400)

            identity, error = await _check_auth(col, OP_READ, request, params, opts)
            if error:
                return error

            # on_pull trigger: sync from primary before serving the local copy
            if (
                opts.replica_manager is not None
                and col.remote is not None
                and SyncTrigger.ON_PULL in col.remote.sync_triggers
            ):
                await opts.replica_manager.on_pull(col.name)

            document_key = _resolve_document_key(col.storage_path, params)
            store = _resolve_store(col, opts.store, params, identity, opts)
            checkpoint_param = request.query_params.get(QUERY_CHECKPOINT)
            is_client_encrypted = bool(col.client_encrypted) or col.encryption == ENCRYPTION_DELEGATED
            return await handle_sync_pull(
                document_key, store, checkpoint_param,
                bool(col.force_full_fetch), is_client_encrypted,
            )

        router.add_api_route(pull_path, pull_handler, methods=["GET"])

    # Push route
    if not col.pull_only:
        push_path = _to_route_path(ACTION_PUSH, col.storage_path)

        rate_limiter = None
        if col.rate_limit and opts.config.rate_limit:
            rate_limiter = RateLimiter(
                window_ms=opts.config.rate_limit.window_ms,
                max_requests=opts.config.rate_limit.max_requests,
            )

        async def push_handler(request: Request, col=col, rate_limiter=rate_limiter) -> JSONResponse:
            params = request.path_params
            if not _validate_all_params(params):
                return JSONResponse({"error": "Invalid path parameter"}, status_code=400)

            identity, error = await _check_auth(col, OP_WRITE, request, params, opts)
            if error:
                return error

            # write_through: block local writes and proxy to primary instead
            if (
                col.remote is not None
                and col.remote.write_mode == WriteMode.WRITE_THROUGH
                and opts.replica_manager is not None
            ):
                return await _proxy_push_to_primary(col, request, opts.replica_manager)

            # pull_only remote: reject local writes
            if col.remote is not None and col.remote.write_mode == WriteMode.PULL_ONLY:
                return JSONResponse(
                    {"error": "This collection is read-only on this server"},
                    status_code=405,
                )

            # Body limit
            content_length = request.headers.get("content-length")
            limit_error = check_body_limit(content_length, col.max_body_bytes)
            if limit_error:
                return limit_error

            # Rate limiting
            if rate_limiter:
                rate_error = rate_limiter.check(identity, request)
                if rate_error:
                    return rate_error

            # Content type check
            content_type = request.headers.get("content-type", "")
            if "application/json" not in content_type:
                return JSONResponse(
                    {"error": "Content-Type must be application/json"},
                    status_code=415,
                )

            body = await request.json()
            if not isinstance(body, dict):
                return JSONResponse({"error": "Body must be a JSON object"}, status_code=400)

            document_key = _resolve_document_key(col.storage_path, params)
            store = _resolve_store(col, opts.store, params, identity, opts)
            is_client_encrypted = bool(col.client_encrypted) or col.encryption == ENCRYPTION_DELEGATED
            response = await handle_sync_push(
                document_key, store, body, identity,
                opts.signature_verifier, is_client_encrypted,
            )

            # Primary-side: notify subscribed replicas after a successful write
            if opts.notification_publisher is not None and response.status_code == 200:
                import json as _json
                resp_body = _json.loads(response.body)
                new_hash = resp_body.get("hash", "")
                timestamp = resp_body.get("timestamp", 0)
                asyncio.create_task(
                    opts.notification_publisher.notify(col.name, new_hash, timestamp)
                )

            return response

        router.add_api_route(push_path, push_handler, methods=["POST"])


def _add_bundled_routes(
    router: APIRouter,
    bundle_name: str,
    collections: list[CollectionConfig],
    opts: SyncRouterOptions,
) -> None:
    storage_path = collections[0].storage_path

    # Pull: combined pull for all collections in the bundle
    pull_path = _to_route_path(ACTION_PULL, storage_path)
    is_any_public = any(ROLE_PUBLIC in c.read_roles for c in collections)

    async def bundle_pull_handler(request: Request) -> JSONResponse:
        params = request.path_params
        if not _validate_all_params(params):
            return JSONResponse({"error": "Invalid path parameter"}, status_code=400)

        if not is_any_public:
            identity, error = await _check_auth(collections[0], OP_READ, request, params, opts)
            if error:
                return error
        else:
            identity = None

        base_key = _resolve_document_key(storage_path, params)
        store = _resolve_store(collections[0], opts.store, params, identity, opts)

        any_client_encrypted = any(
            c.client_encrypted or c.encryption == ENCRYPTION_DELEGATED
            for c in collections
        )
        checkpoint_param = request.query_params.get(QUERY_CHECKPOINT)
        checkpoint = 0
        if not any_client_encrypted and checkpoint_param is not None:
            try:
                parsed = int(checkpoint_param)
            except ValueError:
                return JSONResponse({"error": "Invalid checkpoint"}, status_code=400)
            if parsed < 0 or str(parsed) != checkpoint_param:
                return JSONResponse({"error": "Invalid checkpoint"}, status_code=400)
            checkpoint = parsed

        result: dict[str, Any] = {}
        latest_timestamp = 0

        for col in collections:
            document_key = f"{base_key}/{col.name}"
            pull_result = await pull(store, document_key, checkpoint)
            result[col.name] = {
                "data": pull_result.data,
                "hash": pull_result.hash,
            }
            if pull_result.timestamp > latest_timestamp:
                latest_timestamp = pull_result.timestamp

        return JSONResponse({"collections": result, "timestamp": latest_timestamp})

    router.add_api_route(pull_path, bundle_pull_handler, methods=["GET"])

    # Push: individual push per collection in the bundle
    for col in collections:
        if col.pull_only:
            continue

        push_path = _to_route_path(ACTION_PUSH, storage_path) + f"/{col.name}"

        rate_limiter = None
        if col.rate_limit and opts.config.rate_limit:
            rate_limiter = RateLimiter(
                window_ms=opts.config.rate_limit.window_ms,
                max_requests=opts.config.rate_limit.max_requests,
            )

        async def bundle_push_handler(
            request: Request, col=col, rate_limiter=rate_limiter,
        ) -> JSONResponse:
            params = request.path_params
            if not _validate_all_params(params):
                return JSONResponse({"error": "Invalid path parameter"}, status_code=400)

            identity, error = await _check_auth(col, OP_WRITE, request, params, opts)
            if error:
                return error

            content_length = request.headers.get("content-length")
            limit_error = check_body_limit(content_length, col.max_body_bytes)
            if limit_error:
                return limit_error

            if rate_limiter:
                rate_error = rate_limiter.check(identity, request)
                if rate_error:
                    return rate_error

            content_type = request.headers.get("content-type", "")
            if "application/json" not in content_type:
                return JSONResponse(
                    {"error": "Content-Type must be application/json"},
                    status_code=415,
                )

            body = await request.json()
            if not isinstance(body, dict):
                return JSONResponse({"error": "Body must be a JSON object"}, status_code=400)

            document_key = f"{_resolve_document_key(storage_path, params)}/{col.name}"
            store = _resolve_store(col, opts.store, params, identity, opts)
            is_client_encrypted = bool(col.client_encrypted) or col.encryption == ENCRYPTION_DELEGATED
            return await handle_sync_push(
                document_key, store, body, identity,
                opts.signature_verifier, is_client_encrypted,
            )

        router.add_api_route(push_path, bundle_push_handler, methods=["POST"])


# ── Wildcard catch-all ────────────────────────────────────────────────────


def _add_wildcard_pull_route(router: APIRouter, opts: SyncRouterOptions) -> None:
    """Register a low-priority ``GET /pull/{path:path}`` route for wildcard replication.

    This route is added *after* all explicit collection routes so FastAPI will only
    invoke it for paths that did not match any configured collection.
    """
    wildcard: WildcardRemoteConfig = opts.config.wildcard_remote  # type: ignore[assignment]
    replica_manager = opts.replica_manager

    async def wildcard_pull_handler(request: Request) -> JSONResponse:
        path: str = request.path_params.get("path", "")

        # Validate every segment against the same safe-param regex used elsewhere
        if not path or any(not validate_path_segment(seg) for seg in path.split("/")):
            return JSONResponse({"error": "Invalid path"}, status_code=400)

        # Replica-side auth gate
        if ROLE_PUBLIC not in wildcard.read_roles:
            try:
                auth = await opts.role_resolver(request)
            except Exception:
                return JSONResponse({"error": "Unauthorized"}, status_code=401)
            if not any(r in auth.roles for r in wildcard.read_roles):
                return JSONResponse({"error": "Forbidden"}, status_code=403)

        # Fetch from primary (or serve cached / negative-cached)
        if replica_manager is not None:
            found = await replica_manager.on_pull_wildcard(path)
            if not found:
                return JSONResponse({"error": "Not found"}, status_code=404)

        checkpoint_param = request.query_params.get(QUERY_CHECKPOINT)
        return await handle_sync_pull(
            path, opts.store, checkpoint_param,
            force_full_fetch=False, is_client_encrypted=False,
        )

    router.add_api_route("/pull/{path:path}", wildcard_pull_handler, methods=["GET"])


# ── Public API ───────────────────────────────────────────────────────────


def create_sync_router(opts: SyncRouterOptions) -> APIRouter:
    """Create a FastAPI APIRouter with sync pull/push routes."""
    router = APIRouter()
    config = opts.config

    # Group bundled collections
    bundles: dict[str, list[CollectionConfig]] = {}
    standalone: list[CollectionConfig] = []

    for col in config.collections:
        if col.bundle:
            bundles.setdefault(col.bundle, []).append(col)
        else:
            standalone.append(col)

    for col in standalone:
        _add_collection_routes(router, col, opts)

    for bundle_name, bundle_collections in bundles.items():
        _add_bundled_routes(router, bundle_name, bundle_collections, opts)

    # Wildcard catch-all must be registered last so explicit routes take priority
    if opts.config.wildcard_remote is not None and opts.replica_manager is not None:
        _add_wildcard_pull_route(router, opts)

    return router

File: servers/ts/src/router/route-builder.ts
import { Hono } from "hono"
import type { MiddlewareHandler, Context } from "hono"
import type { IObjectStore } from "../interfaces.js"
import type { SyncConfig, CollectionConfig } from "../config/schema.js"
import { EncryptedObjectStore } from "../encryption/encrypted-store.js"
import { pull } from "../protocol/index.js"
import { handleSyncPull, handleSyncPush, validatePathSegment } from "./helpers.js"
import type { SignatureVerifier } from "./helpers.js"
import { bodyLimit, rateLimitMiddleware } from "./middleware.js"
import {
  ROLE_PUBLIC,
  ROLE_SELF,
  OP_READ,
  OP_WRITE,
  ENCRYPTION_IDENTITY,
  ENCRYPTION_SERVER,
  ENCRYPTION_DELEGATED,
  ACTION_PULL,
  ACTION_PUSH,
  IDENTITY_PARAM,
  IDENTITY_KEY,
  QUERY_CHECKPOINT,
  HKDF_INFO_IDENTITY,
  HKDF_INFO_SERVER,
  HKDF_INFO_DELEGATED,
  HEADER_ENCRYPTION_SECRET,
  HEADER_ENCRYPTION_SALT,
} from "../constants.js"
import type { AccessOperation } from "../constants.js"

// ── Types ────────────────────────────────────────────────────────────────

export interface AuthResult {
  identity: string
  roles: string[]
}

export type RoleResolver = (req: Request) => Promise<AuthResult>

export type RoleEnricher = (
  auth: AuthResult,
  params: Record<string, string>
) => Promise<string[]>

export interface SyncRouterOptions {
  store: IObjectStore
  config: SyncConfig
  roleResolver: RoleResolver
  roleEnricher?: RoleEnricher
  encryptionSecret?: string
  serverEncryptionSecret?: string
  serverIdentity?: string
  identityEncryptionInfo?: string
  serverEncryptionInfo?: string
  signatureVerifier?: SignatureVerifier
}

// ── Helpers ──────────────────────────────────────────────────────────────

/**
 * Convert config storagePath template to a Hono route path.
 * "users/{identity}/invoices" -> "/users/:identity/invoices"
 * Prefixed with /pull or /push.
 */
function toRoutePath(action: string, storagePath: string): string {
  const honoParts = storagePath.replace(/\{(\w+)\}/g, ":$1")
  return `/${action}/${honoParts}`
}

/**
 * Resolve a storage path template into a document key.
 * "users/{identity}/invoices" + { identity: "abc" } -> "users/abc/invoices"
 */
function resolveDocumentKey(template: string, params: Record<string, string>): string {
  return template.replace(/\{(\w+)\}/g, (_, key: string) => params[key] ?? "")
}

function validateAllParams(params: Record<string, string>): boolean {
  for (const value of Object.values(params)) {
    if (!validatePathSegment(value)) return false
  }
  return true
}

/**
 * Compose an array of middleware + handler into a single handler.
 */
function compose(
  middlewares: MiddlewareHandler[],
  handler: (c: Context) => Promise<Response>,
): MiddlewareHandler {
  return async (c, _next) => {
    let idx = 0
    const run = async (): Promise<void> => {
      if (idx < middlewares.length) {
        const mw = middlewares[idx++]!
        const result = await mw(c, run)
        if (result instanceof Response) {
          c.res = result
        }
      } else {
        c.res = await handler(c)
      }
    }
    await run()
  }
}

// ── Auth middleware ───────────────────────────────────────────────────────

/**
 * Build auth middleware for a given collection and operation (read/write).
 * Handles: public access, role resolution, "self" auto-grant, role enrichment.
 */
function buildAuthMiddleware(
  col: CollectionConfig,
  operation: AccessOperation,
  opts: SyncRouterOptions,
): MiddlewareHandler | null {
  const requiredRoles = operation === OP_READ ? col.readRoles : col.writeRoles

  // Public access — no auth needed
  if (requiredRoles.includes(ROLE_PUBLIC)) return null

  return async (c, next) => {
    let auth: AuthResult
    try {
      auth = await opts.roleResolver(c.req.raw)
    } catch {
      return c.json({ error: "Unauthorized" }, 401)
    }

    // Store identity in context for downstream use (rate limiting, encryption)
    c.set(IDENTITY_KEY, auth.identity)

    const effectiveRoles = new Set(auth.roles)

    // Auto-grant "self" when {identity} in path matches authenticated identity
    if (col.storagePath.includes(IDENTITY_PARAM)) {
      const params = c.req.param() as Record<string, string>
      if (params[IDENTITY_KEY] === auth.identity) {
        effectiveRoles.add(ROLE_SELF)
      }
    }

    // Enrich roles with context-dependent roles (e.g. "owner")
    if (opts.roleEnricher) {
      const params = c.req.param() as Record<string, string>
      const extra = await opts.roleEnricher(auth, params)
      for (const role of extra) effectiveRoles.add(role)
    }

    // Check if any effective role matches the required roles
    const hasAccess = requiredRoles.some(r => effectiveRoles.has(r))
    if (!hasAccess) {
      return c.json({ error: "Forbidden" }, 403)
    }

    await next()
  }
}

// ── Store resolution ─────────────────────────────────────────────────────

function resolveStore(
  col: CollectionConfig,
  baseStore: IObjectStore,
  params: Record<string, string>,
  identity: string | undefined,
  opts: SyncRouterOptions,
  req?: Request,
): IObjectStore | Response {
  if (col.encryption === ENCRYPTION_IDENTITY) {
    if (!opts.encryptionSecret) throw new Error(`Collection "${col.name}" requires encryptionSecret`)
    const salt = identity ?? params[IDENTITY_KEY] ?? ""
    return new EncryptedObjectStore(
      baseStore,
      opts.encryptionSecret,
      salt,
      opts.identityEncryptionInfo ?? HKDF_INFO_IDENTITY,
    )
  }
  if (col.encryption === ENCRYPTION_SERVER) {
    if (!opts.serverEncryptionSecret) throw new Error(`Collection "${col.name}" requires serverEncryptionSecret`)
    if (!opts.serverIdentity) throw new Error(`Collection "${col.name}" requires serverIdentity`)
    return new EncryptedObjectStore(
      baseStore,
      opts.serverEncryptionSecret,
      opts.serverIdentity,
      opts.serverEncryptionInfo ?? HKDF_INFO_SERVER,
    )
  }
  if (col.encryption === ENCRYPTION_DELEGATED) {
    const secret = req?.headers.get(HEADER_ENCRYPTION_SECRET)
    const salt = req?.headers.get(HEADER_ENCRYPTION_SALT)
    if (!secret || !salt) {
      return Response.json(
        { error: "Delegated encryption requires X-Encryption-Secret and X-Encryption-Salt headers" },
        { status: 400 },
      )
    }
    return new EncryptedObjectStore(
      baseStore,
      secret,
      salt,
      HKDF_INFO_DELEGATED,
    )
  }
  return baseStore
}

// ── Route building ───────────────────────────────────────────────────────

function addCollectionRoutes(router: Hono, col: CollectionConfig, opts: SyncRouterOptions): void {
  // Pull route
  if (!col.pushOnly) {
    const pullPath = toRoutePath(ACTION_PULL, col.storagePath)
    const authMw = buildAuthMiddleware(col, OP_READ, opts)
    const middlewares: MiddlewareHandler[] = authMw ? [authMw] : []

    const handler = async (c: Context) => {
      const params = c.req.param() as Record<string, string>
      if (!validateAllParams(params)) {
        return c.json({ error: "Invalid path parameter" }, 400)
      }
      const documentKey = resolveDocumentKey(col.storagePath, params)
      const identity = (c.get(IDENTITY_KEY) as string | undefined)
      const storeOrError = resolveStore(col, opts.store, params, identity, opts, c.req.raw)
      if (storeOrError instanceof Response) return storeOrError
      return handleSyncPull(c, documentKey, storeOrError, col.forceFullFetch, col.clientEncrypted)
    }

    router.get(pullPath, compose(middlewares, handler))
  }

  // Push route
  if (!col.pullOnly) {
    const pushPath = toRoutePath(ACTION_PUSH, col.storagePath)
    const middlewares: MiddlewareHandler[] = []

    const authMw = buildAuthMiddleware(col, OP_WRITE, opts)
    if (authMw) middlewares.push(authMw)
    middlewares.push(bodyLimit(col.maxBodyBytes))
    if (col.rateLimit && opts.config.rateLimit) {
      middlewares.push(rateLimitMiddleware(opts.config.rateLimit))
    }

    const handler = async (c: Context) => {
      const params = c.req.param() as Record<string, string>
      if (!validateAllParams(params)) {
        return c.json({ error: "Invalid path parameter" }, 400)
      }
      const documentKey = resolveDocumentKey(col.storagePath, params)
      const identity = (c.get(IDENTITY_KEY) as string | undefined)
      const storeOrError = resolveStore(col, opts.store, params, identity, opts, c.req.raw)
      if (storeOrError instanceof Response) return storeOrError
      return handleSyncPush(c, documentKey, storeOrError, identity, opts.signatureVerifier, col.clientEncrypted)
    }

    router.post(pushPath, compose(middlewares, handler))
  }
}

function addBundledRoutes(router: Hono, bundleName: string, collections: CollectionConfig[], opts: SyncRouterOptions): void {
  const storagePath = collections[0]!.storagePath

  // Pull: combined pull for all collections in the bundle
  const pullPath = toRoutePath(ACTION_PULL, storagePath)
  const isAnyPublic = collections.some(c => c.readRoles.includes(ROLE_PUBLIC))
  const pullAuthMw = isAnyPublic ? null : buildAuthMiddleware(collections[0]!, OP_READ, opts)
  const pullMiddlewares: MiddlewareHandler[] = pullAuthMw ? [pullAuthMw] : []

  const pullHandler = async (c: Context) => {
    const params = c.req.param() as Record<string, string>
    if (!validateAllParams(params)) {
      return c.json({ error: "Invalid path parameter" }, 400)
    }
    const baseKey = resolveDocumentKey(storagePath, params)
    const identity = (c.get(IDENTITY_KEY) as string | undefined)
    const storeOrError = resolveStore(collections[0]!, opts.store, params, identity, opts, c.req.raw)
    if (storeOrError instanceof Response) return storeOrError
    const store = storeOrError

    const anyClientEncrypted = collections.some(c => c.clientEncrypted)
    const checkpointParam = c.req.query(QUERY_CHECKPOINT)
    let checkpoint = 0
    if (!anyClientEncrypted && checkpointParam !== undefined) {
      const parsed = parseInt(checkpointParam, 10)
      if (!Number.isInteger(parsed) || parsed < 0 || String(parsed) !== checkpointParam) {
        return c.json({ error: "Invalid checkpoint" }, 400)
      }
      checkpoint = parsed
    }

    // Pull each collection as a separate document under the bundle key
    const result: Record<string, unknown> = {}
    let latestTimestamp = 0

    for (const col of collections) {
      const documentKey = `${baseKey}/${col.name}`
      const pullResult = await pull(store, documentKey, checkpoint)
      result[col.name] = {
        data: pullResult.data,
        hash: pullResult.hash,
      }
      if (pullResult.timestamp > latestTimestamp) latestTimestamp = pullResult.timestamp
    }

    return c.json({ collections: result, timestamp: latestTimestamp })
  }

  router.get(pullPath, compose(pullMiddlewares, pullHandler))

  // Push: individual push per collection in the bundle
  for (const col of collections) {
    if (col.pullOnly) continue

    const pushPath = toRoutePath(ACTION_PUSH, storagePath) + `/${col.name}`
    const middlewares: MiddlewareHandler[] = []
    const authMw = buildAuthMiddleware(col, OP_WRITE, opts)
    if (authMw) middlewares.push(authMw)
    middlewares.push(bodyLimit(col.maxBodyBytes))
    if (col.rateLimit && opts.config.rateLimit) {
      middlewares.push(rateLimitMiddleware(opts.config.rateLimit))
    }

    const handler = async (c: Context) => {
      const params = c.req.param() as Record<string, string>
      if (!validateAllParams(params)) {
        return c.json({ error: "Invalid path parameter" }, 400)
      }
      const documentKey = `${resolveDocumentKey(storagePath, params)}/${col.name}`
      const identity = (c.get(IDENTITY_KEY) as string | undefined)
      const storeOrError = resolveStore(col, opts.store, params, identity, opts, c.req.raw)
      if (storeOrError instanceof Response) return storeOrError
      return handleSyncPush(c, documentKey, storeOrError, identity, opts.signatureVerifier, col.clientEncrypted)
    }

    router.post(pushPath, compose(middlewares, handler))
  }
}

// ── Public API ───────────────────────────────────────────────────────────

export function createSyncRouter(opts: SyncRouterOptions): Hono {
  const router = new Hono()
  const { config } = opts

  // Group bundled collections
  const bundles = new Map<string, CollectionConfig[]>()
  const standalone: CollectionConfig[] = []

  for (const col of config.collections) {
    if (col.bundle) {
      const list = bundles.get(col.bundle) ?? []
      list.push(col)
      bundles.set(col.bundle, list)
    } else {
      standalone.push(col)
    }
  }

  for (const col of standalone) {
    addCollectionRoutes(router, col, opts)
  }

  for (const [bundleName, collections] of bundles) {
    addBundledRoutes(router, bundleName, collections, opts)
  }

  return router
}
